_drafts/reading-legacy-code-to-harmonize-ruby-and-rails-code-loaders.md

```diff
@@ -0,0 +1,701 @@
+no tests; there were no docs; and the specifications were locked in people heads as tribal knowledge.  I also had to read `ActiveSupport::Dependencies` to understand what hidden preconditions it had on code it loaded.  Integrating this knowledge I was able to make the Metasploit Module loader compatible by dynamically assigning deterministic namespace names to the loaded Metasploit Modules.
+
+* Table of Contents
+{:toc}
+
+# Background
+
+## Code-Loading in Ruby
+
+Ruby has 3 ways to load code in the standard library:
+
+1. `load`
+2. `require`
+3. `autoload`
+
+All three search for files in the load path, which is an array of String representing directories under which relative paths can be resolved.  (If an absolute path is given the it is loaded directly without checking under directories in the load path.) The current load path can be accessed in the `$:` or the less cryptic, `$LOAD_PATH`.
+
+### `load`
+
+`load` is the lowest level of the code loading methods.  It will resolve relative paths in $LOAD_PATH, but it won't automatically add on the `.rb` extension for you (or .dll, .so, or .dylib extension for C extensions).  `load` will also reload files that have already been loaded, which is a good thing if you want to reload code in development, but a bad thing if you're trying to be efficient when starting up a slow application that has multiple ways for a file to be required.
+
+### `require`
+
+`require` is a little higher level than `load`.  It will automatically add on the `.rb` extension and any extensions your system uses for C extension like `.dll` on Windows, `.so` on Linux, or `.dylib` on OS X.  This has the benefit of your code not needing to know whether you're using a pure Ruby library with a `.rb` extension or C extension library with an different extension.  This power of `require` is what allows you to use C-extension gems in your bundle without having to change all your requires.  `require` also has the nice benefit that it's thread-safe due a lock used in the Ruby VM while `load` is not thread-safe.
+
+### `autoload`
+
+`load` and `require` both have the downside that as soon as they are encountered, they load the code, which means you have to wait for all the libraries you might use, but may not use for a given run of your program, to load.  If you want to only load the code you actually use, you can use `autoload`, which takes a relative constant name, that is, a Class or Module name and the path to the file to load to define that constant:
+
+```ruby
+# lib/parent.rb
+module Parent
+  autoload :Child, 'parent/child'
+end
+
+# lib/parent/child.rb
+module Parent::Child
+end
+```
+
+When `Parent::Child` is accessed for the first time, `parent/child` will be loaded.  In 1.9.3, this loading will happen once, but is not thread-safe; in 2.0+ this loading will happen once, but is thread-safe and can be thought of as just an on-use `require` as `autoload` uses the same VM lock as `require` in MRI Ruby 2.0+.
```

Comment:
When `Parent::Child` is accessed for the first time, `parent/child` will be loaded.  In **Ruby** 1.9.3, this loading will happen once**~~,~~ (remove comma)** but is not thread-safe; in **Ruby** 2.0+ this loading will happen once**~~,~~ (remove comma)** but ***is*** **(add emphasis)** thread-safe and can be thought of as just an on-use `require` as `autoload` uses the same VM lock as `require` in MRI Ruby 2.0+.

Perhaps the phrase "lazy loading" would help drive the point home.